File: advanced/byosnap-node-ts/src/controllers/charactersController.ts
import { ErrorResponse } from '../types/responses';
import { CharactersResponse } from '../models/charactersModel';
import { authMiddleware } from '../middleware/authMiddleware';
import { Request as ExpressRequest } from 'express';
import { Controller, Route, Get, Path, Extension, Middlewares, TsoaResponse, Response, Res, Request } from 'tsoa';

// # @GOTCHAS 👋 - Please read GOTCHAS.md

// ===========================================================================
// Regular API Endpoints exposed by the Snap
// ===========================================================================

@Route('/v1/byosnap-advanced/users')
export class CharactersController extends Controller {

    /**
     * @summary Character APIs
     */
    @Get("{userId}/characters/active")
    @Extension("x-description", 'Get active characters for a user. This API supports User, API-Key, and Internal auth types.')
    @Extension("x-snapser-auth-types", ["user", "api-key", "internal"])
    @Response<CharactersResponse>(200, "Characters retrieved successfully")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["user", "api-key", "internal"])])
    public async getActiveCharacters(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<CharactersResponse> {
        return {
            characters: [],
        };
    }
}

File: advanced/ByoSnapCSharp/Controllers/CharactersController.cs
using Microsoft.AspNetCore.Mvc;
using ByoSnapCSharp.Filters;
using ByoSnapCSharp.Models;
using ByoSnapCSharp.Utilities;
using Swashbuckle.AspNetCore.Annotations;
using System.Collections.Generic;

namespace ByoSnapCSharp.Controllers
{
  // Regular API Endpoints exposed by the Snap

  [ApiController]
  [Route("v1/byosnap-advanced/users/{UserId}")]
  [Produces("application/json")]
  public class CharactersController : ControllerBase
  {
    [HttpGet("characters/active")]
    [SnapserAuth(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(CharactersResponseSchema))]
    [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ErrorResponseSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "GetActiveCharacters",
      Summary = "Character APIs",
      Description = "Get active characters for a user. This API supports User, API-Key, and Internal auth types."
    )]
    public ActionResult<CharactersResponseSchema> GetActiveCharacters([FromRoute] UserIdParameterSchema userParams)
    {
      return Ok(new CharactersResponseSchema
      {
        Characters = new List<string>()
      });
    }
  }
}
